File: python/dvfc/metricflow.py
"""
Compile dbt_metric → SQL.

Order:
1. Fixture files (semantic/<metric>.sql, metrics/<metric>.sql)
2. In-process MetricFlow Python API (optional extra `dvfc[metricflow]`)
3. CLI: `mf query --explain` or `dbt sl query --compile`
"""

from __future__ import annotations

import os
import re
import shutil
import subprocess
from pathlib import Path
from typing import List, Literal, Optional, Sequence, Tuple

Source = Literal["fixture", "metricflow-python", "metricflow-cli"]


def find_fixture_sql(
    metric: str,
    *,
    spec_dir: Path,
    project_root: Optional[Path] = None,
) -> Optional[str]:
    root = project_root or spec_dir
    candidates = [
        spec_dir / "semantic" / f"{metric}.sql",
        spec_dir / "metrics" / f"{metric}.sql",
        root / "semantic" / f"{metric}.sql",
        spec_dir / "dbt-stub" / "semantic" / f"{metric}.sql",
    ]
    for p in candidates:
        if p.is_file():
            return p.read_text(encoding="utf-8")
    return None


def find_dbt_project(start: Path) -> Optional[Path]:
    cur = start.resolve()
    for _ in range(12):
        if (cur / "dbt_project.yml").is_file():
            return cur
        if cur.parent == cur:
            break
        cur = cur.parent
    return None


def extract_sql_from_cli_output(stdout: str) -> str:
    cleaned_lines = []
    for line in stdout.splitlines():
        t = line.strip()
        if not t:
            cleaned_lines.append(line)
            continue
        if t[0] in "⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏":
            continue
        if t.lower().startswith("initiating query"):
            continue
        if t.startswith("🔎"):
            continue
        if t.lower().startswith("sql:"):
            continue
        cleaned_lines.append(line)
    cleaned = "\n".join(cleaned_lines)
    fence = re.search(r"```(?:sql)?\s*([\s\S]*?)```", cleaned, re.I)
    if fence and fence.group(1).strip():
        return fence.group(1).strip()
    m = re.search(r"\bselect\b", cleaned, re.I)
    if m:
        sql = cleaned[m.start() :].strip()
        sql = re.sub(r"\n(?:Query|Returned|Success|Error).*$", "", sql, flags=re.I | re.S).strip()
        if len(sql) > 10:
            return sql
    raise RuntimeError("Could not parse SQL from MetricFlow CLI output")


def _compile_via_python_api(
    metric: str,
    group_by: Sequence[str],
    where: Optional[str],
    project_dir: Path,
) -> Optional[str]:
    """
    Best-effort in-process MetricFlow. Returns None if API unavailable.
    """
    try:
        # dbt-metricflow / metricflow layouts vary by version — try common entry points
        try:
            from metricflow.engine.metricflow_engine import (  # type: ignore
                MetricFlowEngine,
                MetricFlowQueryRequest,
            )
        except ImportError:
            from metricflow import MetricFlowEngine, MetricFlowQueryRequest  # type: ignore
    except ImportError:
        return None

    try:
        # Many installs need a configured dbt project / semantic manifest already parsed
        engine = MetricFlowEngine.create_from_dbt_project_root(str(project_dir))  # type: ignore[attr-defined]
    except Exception:
        try:
            from metricflow.engine.metricflow_engine import MetricFlowEngine as MFE  # type: ignore

            engine = MFE()  # type: ignore
        except Exception:
            return None

    try:
        request = MetricFlowQueryRequest.create_with_names(  # type: ignore[attr-defined]
            metric_names=[metric],
            group_by_names=list(group_by) or ["metric_time"],
            where_constraints=[where] if where else None,
        )
        result = engine.explain(request)  # type: ignore[attr-defined]
        sql = getattr(result, "sql", None) or getattr(result, "rendered_sql", None)
        if sql:
            return str(sql)
        # Some versions return explain plan object
        plan = getattr(result, "sql_statement", None)
        if plan:
            return str(plan)
    except Exception:
        return None
    return None


def _compile_via_cli(
    metric: str,
    group_by: Sequence[str],
    where: Optional[str],
    project_dir: Path,
) -> str:
    mode = (os.environ.get("DVFC_METRICFLOW_MODE") or "auto").lower()
    bin_override = os.environ.get("DVFC_METRICFLOW_BIN")

    if mode in ("dbt-sl", "dbt_sl", "sl") or (bin_override and "dbt" in bin_override):
        binary = bin_override or "dbt"
        args = [
            "sl",
            "query",
            "--metrics",
            metric,
            "--group-by",
            ",".join(group_by) or "metric_time",
            "--compile",
        ]
    else:
        binary = bin_override or shutil.which("mf") or "mf"
        args = [
            "query",
            "--metrics",
            metric,
            "--group-by",
            ",".join(group_by) or "metric_time",
            "--explain",
            "--quiet",
        ]
    if where:
        args.extend(["--where", where])

    def run(cmd_args: List[str]) -> str:
        proc = subprocess.run(
            [binary, *cmd_args],
            cwd=str(project_dir),
            capture_output=True,
            text=True,
            timeout=120,
        )
        out = (proc.stdout or "") + "\n" + (proc.stderr or "")
        if proc.returncode != 0 and not re.search(r"\bselect\b", out, re.I):
            raise RuntimeError(out.strip() or f"{binary} exited {proc.returncode}")
        return extract_sql_from_cli_output(out)

    try:
        return run(args)
    except Exception:
        if "--quiet" in args:
            return run([a for a in args if a != "--quiet"])
        raise


def compile_dbt_metric(
    metric: str,
    *,
    group_by: Optional[Sequence[str]] = None,
    where: Optional[str] = None,
    spec_dir: Optional[Path] = None,
    project_root: Optional[Path] = None,
    dbt_project: Optional[Path] = None,
    skip_invoke: bool = False,
    cache: bool = False,
) -> Tuple[str, Source]:
    """
    Resolve metric → SQL. Returns (sql, source).
    """
    spec = (spec_dir or Path.cwd()).resolve()
    root = (project_root or spec).resolve()
    gb = list(group_by) if group_by else ["metric_time"]

    fixture = find_fixture_sql(metric, spec_dir=spec, project_root=root)
    if fixture:
        return fixture, "fixture"

    if skip_invoke or os.environ.get("DVFC_METRICFLOW_SKIP") == "1":
        raise RuntimeError(
            f"dbt_metric '{metric}' has no fixture and invoke is skipped"
        )

    project = (
        dbt_project
        or (Path(os.environ["DVFC_DBT_PROJECT"]) if os.environ.get("DVFC_DBT_PROJECT") else None)
        or find_dbt_project(spec)
        or find_dbt_project(root)
        or root
    )

    sql = _compile_via_python_api(metric, gb, where, project)
    source: Source
    if sql:
        source = "metricflow-python"
    else:
        sql = _compile_via_cli(metric, gb, where, project)
        source = "metricflow-cli"

    if cache or os.environ.get("DVFC_METRICFLOW_CACHE") == "1":
        out = spec / "semantic" / f"{metric}.sql"
        out.parent.mkdir(parents=True, exist_ok=True)
        out.write_text(f"-- Compiled for metric {metric} via {source}\n{sql}\n", encoding="utf-8")

    return sql, source

File: packages/core/src/metricflow.ts
/**
 * MetricFlow / dbt semantic layer: compile metric → SQL
 *
 * Resolution order for data.type: dbt_metric:
 * 1. Checked-in fixture: semantic/<metric>.sql or metrics/<metric>.sql
 * 2. Live invoke: `mf query --explain` or `dbt sl query --compile`
 *
 * Env:
 *   DVFC_METRICFLOW_BIN   — override binary (default: auto-detect mf / dbt)
 *   DVFC_METRICFLOW_MODE  — mf | dbt-sl | auto (default auto)
 *   DVFC_DBT_PROJECT      — dbt project root (else walk parents for dbt_project.yml)
 *   DVFC_METRICFLOW_CACHE — if "1", write compiled SQL to semantic/<metric>.sql
 *   DVFC_METRICFLOW_SKIP  — if "1", skip live invoke (fixtures only)
 */

import { access, mkdir, readFile, writeFile } from 'fs/promises';
import { dirname, join, resolve as resolvePath } from 'path';
import { execFile } from 'child_process';
import { promisify } from 'util';
import type { DataRefDbtMetric } from './ir.js';

const execFileAsync = promisify(execFile);

export interface MetricFlowCompileOptions {
  specDir: string;
  projectRoot?: string;
  /** Override cwd for mf/dbt (dbt project root) */
  dbtProjectDir?: string;
  /** Skip live CLI invoke */
  skipInvoke?: boolean;
  /** Cache compiled SQL under semantic/ */
  cache?: boolean;
}

async function fileExists(p: string): Promise<boolean> {
  try {
    await access(p);
    return true;
  } catch {
    return false;
  }
}

/** Locate checked-in compiled SQL for a metric */
export async function findMetricFixtureSql(
  metric: string,
  opts: { specDir: string; projectRoot?: string; dbtStubDir?: string }
): Promise<string | null> {
  const dbtStub = opts.dbtStubDir ?? join(opts.specDir, 'dbt-stub');
  const candidates = [
    join(opts.specDir, 'semantic', `${metric}.sql`),
    join(opts.specDir, 'metrics', `${metric}.sql`),
    join(opts.projectRoot ?? opts.specDir, 'semantic', `${metric}.sql`),
    join(dbtStub, 'semantic', `${metric}.sql`),
  ];
  for (const p of candidates) {
    if (await fileExists(p)) {
      return readFile(p, 'utf-8');
    }
  }
  return null;
}

/** Walk up from start looking for dbt_project.yml */
export async function findDbtProjectRoot(start: string): Promise<string | null> {
  let dir = resolvePath(start);
  for (let i = 0; i < 12; i++) {
    if (await fileExists(join(dir, 'dbt_project.yml'))) return dir;
    const parent = dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return null;
}

function detectMode(): 'mf' | 'dbt-sl' {
  const mode = (process.env.DVFC_METRICFLOW_MODE || 'auto').toLowerCase();
  if (mode === 'mf' || mode === 'metricflow') return 'mf';
  if (mode === 'dbt-sl' || mode === 'dbt_sl' || mode === 'sl') return 'dbt-sl';
  // auto: prefer DVFC_METRICFLOW_BIN hint, else mf
  const bin = process.env.DVFC_METRICFLOW_BIN || '';
  if (bin.includes('dbt')) return 'dbt-sl';
  return 'mf';
}

/**
 * Extract SQL from MetricFlow / dbt sl explain/compile output.
 * Strips spinner lines and takes the largest SELECT-looking block.
 */
export function extractSqlFromMetricFlowOutput(stdout: string): string {
  const cleaned = stdout
    .split('\n')
    .filter((line) => {
      const t = line.trim();
      if (!t) return true;
      if (/^[⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏]/.test(t)) return false;
      if (/^initiating query/i.test(t)) return false;
      if (/^🔎/.test(t)) return false;
      if (/^sql:/i.test(t)) return false;
      return true;
    })
    .join('\n');

  // Prefer fenced sql blocks
  const fence = cleaned.match(/```(?:sql)?\s*([\s\S]*?)```/i);
  if (fence?.[1]?.trim()) return fence[1].trim();

  // Find SELECT … to end (or first blank line after substantial SQL)
  const selectIdx = cleaned.search(/\bselect\b/i);
  if (selectIdx >= 0) {
    let sql = cleaned.slice(selectIdx).trim();
    // Cut trailing status lines
    sql = sql.replace(/\n(?:Query|Returned|Success|Error).*$/is, '').trim();
        if (sql.length > 10) return sql;
  }

  throw new Error(
    'Could not parse SQL from MetricFlow output. Use --explain/--compile and ensure mf/dbt sl is configured.'
  );
}

function buildArgs(
  mode: 'mf' | 'dbt-sl',
  ref: DataRefDbtMetric
): { bin: string; args: string[] } {
  const metrics = ref.metric;
  const groupBy = (ref.group_by ?? ['metric_time']).join(',');
  const where = ref.where;

  if (mode === 'dbt-sl') {
    const bin = process.env.DVFC_METRICFLOW_BIN || 'dbt';
    const args = [
      'sl',
      'query',
      '--metrics',
      metrics,
      '--group-by',
      groupBy,
      '--compile',
    ];
    if (where) args.push('--where', where);
    return { bin, args };
  }

  const bin = process.env.DVFC_METRICFLOW_BIN || 'mf';
  const args = [
    'query',
    '--metrics',
    metrics,
    '--group-by',
    groupBy,
    '--explain',
  ];
  if (where) args.push('--where', where);
  // Prefer quiet if available (MetricFlow ≥ Feb 2025)
  args.push('--quiet');
  return { bin, args };
}

/**
 * Invoke MetricFlow CLI to compile metric SQL.
 */
export async function invokeMetricFlow(
  ref: DataRefDbtMetric,
  opts: MetricFlowCompileOptions
): Promise<string> {
  if (opts.skipInvoke || process.env.DVFC_METRICFLOW_SKIP === '1') {
    throw new Error('MetricFlow invoke skipped (DVFC_METRICFLOW_SKIP or skipInvoke)');
  }

  const cwd =
    opts.dbtProjectDir ||
    process.env.DVFC_DBT_PROJECT ||
    (await findDbtProjectRoot(opts.specDir)) ||
    (await findDbtProjectRoot(opts.projectRoot ?? opts.specDir)) ||
    opts.projectRoot ||
    opts.specDir;

  const mode = detectMode();
  const { bin, args } = buildArgs(mode, ref);

  let stdout: string;
  let stderr: string;
  try {
    const result = await execFileAsync(bin, args, {
      cwd,
      env: process.env,
      maxBuffer: 10 * 1024 * 1024,
      timeout: 120_000,
    });
    stdout = result.stdout || '';
    stderr = result.stderr || '';
  } catch (err: unknown) {
    const e = err as { stdout?: string; stderr?: string; message?: string };
    // Some mf versions write SQL to stdout even on non-zero; try parse
    if (e.stdout && /\bselect\b/i.test(e.stdout)) {
      stdout = e.stdout;
      stderr = e.stderr || '';
    } else {
      // Retry without --quiet if unsupported
      if (args.includes('--quiet')) {
        const retryArgs = args.filter((a) => a !== '--quiet');
        try {
          const result = await execFileAsync(bin, retryArgs, {
            cwd,
            env: process.env,
            maxBuffer: 10 * 1024 * 1024,
            timeout: 120_000,
          });
          stdout = result.stdout || '';
          stderr = result.stderr || '';
        } catch (err2: unknown) {
          const e2 = err2 as { stdout?: string; stderr?: string; message?: string };
          throw new Error(
            `MetricFlow invoke failed (${bin} ${retryArgs.join(' ')}) in ${cwd}:\n` +
              `${e2.stderr || e2.stdout || e2.message || String(err2)}`
          );
        }
      } else {
        throw new Error(
          `MetricFlow invoke failed (${bin} ${args.join(' ')}) in ${cwd}:\n` +
            `${e.stderr || e.stdout || e.message || String(err)}`
        );
      }
    }
  }

  const combined = `${stdout}\n${stderr}`;
  return extractSqlFromMetricFlowOutput(combined);
}

/**
 * Resolve dbt_metric → SQL string (fixture first, then live MetricFlow).
 */
export async function compileDbtMetricSql(
  ref: DataRefDbtMetric,
  opts: MetricFlowCompileOptions & { dbtStubDir?: string }
): Promise<{ sql: string; source: 'fixture' | 'metricflow' }> {
  const fixture = await findMetricFixtureSql(ref.metric, opts);
  if (fixture) {
    return { sql: fixture, source: 'fixture' };
  }

  try {
    const sql = await invokeMetricFlow(ref, opts);
    const cache =
      opts.cache === true || process.env.DVFC_METRICFLOW_CACHE === '1';
    if (cache) {
      const out = join(opts.specDir, 'semantic', `${ref.metric}.sql`);
      await mkdir(dirname(out), { recursive: true });
      await writeFile(
        out,
        `-- Compiled by MetricFlow for metric ${ref.metric}\n${sql}\n`
      );
    }
    return { sql, source: 'metricflow' };
  } catch (err) {
    throw new Error(
      `dbt_metric '${ref.metric}' not resolved.\n` +
        `  Tried fixtures under semantic/${ref.metric}.sql and metrics/${ref.metric}.sql\n` +
        `  MetricFlow: ${err instanceof Error ? err.message : String(err)}\n` +
        `  Tip: run \`dbt parse\` then \`mf query --metrics ${ref.metric} --group-by metric_time --explain\`, ` +
        `or commit compiled SQL under semantic/. See docs/dbt-metrics.md.`
    );
  }
}
